### PR TITLE
iox-#743 Validate callbacks in c example

### PR DESCRIPTION
**Pull request by FerdinandSpitzschnueffler**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Partially closes #743

### CODE REVIEW

File: iceoryx_examples/callbacks_in_c/README.md [hunk 1]
```diff
@@ -126,8 +136,11 @@ iox_listener_detach_subscriber_event(listener, subscriberRight, SubscriberEvent_
 
 In a last step we have to release all acquired resources
 
+<!--[geoffrey][iceoryx_examples/callbacks_in_c/ice_c_callbacks_subscriber.c][cleanup]-->
 ```c
+#if !defined(_WIN32)
 pthread_join(heartbeatTriggerThread, NULL);
+#endif
```
Comment (elBoberido) [inline iceoryx_examples/callbacks_in_c/README.md:143]: I wonder whether this 3 lines could be removed without reducing the information for the user.

**Review by elBoberido — APPROVED** [review]

File: iceoryx_examples/callbacks_in_c/README.md [hunk 2]
```diff
@@ -12,7 +12,7 @@ one accessing it or that it is accessed with a guard like a `mutex`.
 
 For a general introduction into the Listener concept please take a look at
 the first part of the
-[Listener C++ example](https://github.com/eclipse-iceoryx/iceoryx/tree/master/iceoryx_examples/callbacks) 
+[Listener C++ example](https://github.com/eclipse-iceoryx/iceoryx/tree/master/iceoryx_examples/callbacks)
```
Comment (mossmaurice) [inline iceoryx_examples/callbacks_in_c/README.md:15]: Same for all the other links, we need to link to a fixed version.

```suggestion
[Listener C++ example](https://github.com/eclipse-iceoryx/iceoryx/tree/v2.0.0/iceoryx_examples/callbacks)
```

File: iceoryx_examples/callbacks_in_c/README.md (hunk 1 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_examples/callbacks_in_c/README.md:143]: I asked myself the same question :smiley:

**Review by elfenpiff — DISMISSED** [review]

File: iceoryx_examples/callbacks_in_c/README.md [hunk 3]
```diff
@@ -170,74 +182,85 @@ void onSampleReceivedCallback(iox_sub_t subscriber)
...
+}
 ```
 
 ### Additional context data for callbacks (ice_c_callbacks_with_context_data.c)
 
-Sometimes you would like to modify data structures within the callback which 
-are not globally available. To facilitate this we implemented functions called 
-`iox_listener_attach_***_event_with_context_data` which are enabling you to 
-to provide an additional void pointer to the callback as second argument.
+Sometimes you would like to modify data structures within the callback which
```
Comment (MatthiasKillat) [inline iceoryx_examples/callbacks_in_c/README.md:216]:
```suggestion
Sometimes we would like to modify data structures which
```

File: iceoryx_examples/callbacks_in_c/README.md [hunk 4]
```diff
@@ -170,74 +182,85 @@ void onSampleReceivedCallback(iox_sub_t subscriber)
...
+are not globally available. To facilitate this we implemented functions called
```
Comment (MatthiasKillat) [inline iceoryx_examples/callbacks_in_c/README.md:217]:
```suggestion
are not globally available within the callback. To facilitate this we provide the functions
```

File: iceoryx_examples/callbacks_in_c/README.md [hunk 5]
```diff
@@ -170,74 +182,85 @@ void onSampleReceivedCallback(iox_sub_t subscriber)
...
+`iox_listener_attach_***_event_with_context_data` which allow you to provide an
```
Comment (MatthiasKillat) [inline iceoryx_examples/callbacks_in_c/README.md:218]:
```suggestion
`iox_listener_attach_***_event_with_context_data` which allows to provide an
```

File: iceoryx_examples/callbacks_in_c/README.md [hunk 6]
```diff
@@ -170,74 +182,85 @@ void onSampleReceivedCallback(iox_sub_t subscriber)
...
-    counterService.leftCache.isSet = false;
-    counterService.rightCache.isSet = false;
+CounterService counterService;
+counterService.leftCache.isSet = false;
+counterService.rightCache.isSet = false;
 ```
 
-The callback gets an additional void pointer argument which we cast then to 
-our CounterService to perform the same tasks as in the previous example but now 
+The callback gets an additional void pointer argument which we cast then to
```
Comment (MatthiasKillat) [inline iceoryx_examples/callbacks_in_c/README.md:234]:
```suggestion
The callback takes an additional void pointer argument which we cast then to
```

**Review by elBoberido — APPROVED** [review]

**Pull request merged**